File: frontend/src/api/browser.ts
import api from './client'

export interface BrowserFile {
  id:            string
  case_id:       string
  filename:      string
  artifact_type: string
  status:        'pending' | 'parsing' | 'ready' | 'error'
  entry_count:   number | null
  error_msg:     string | null
  uploaded_at:   string
  parsed_at:     string | null
  added_to_evidence:      boolean
  parse_progress:         number
  parse_duration_seconds: number | null
  /** Original CSV column names — populated after parse completes. */
  columns:       string[]
}

export interface BrowserEntry {
  row_num:        number
  artifact_type:  string
  event_timestamp: string | null
  url:            string | null
  title:          string | null
  browser:        string | null
  profile:        string | null
  username:       string | null
  /** Every original CSV column → value (empty values omitted). */
  raw_data:       Record<string, string>
}

/** A BrowserEntry pinned to the right-panel selection, enriched with file metadata. */
export interface PinnedBrowserEntry extends BrowserEntry {
  _key:      string   // `${fileId}:${row_num}`
  _fileId:   string
  _filename: string
}

export interface BrowserEntriesPage {
  total:     number
  page:      number
  page_size: number
  pages:     number
  items:     BrowserEntry[]
}

export interface BrowserSummary {
  total_entries:    number
  artifact_type:    string
  oldest_timestamp: string | null
  newest_timestamp: string | null
  top_browsers:     Array<{ browser: string; count: number }>
  top_domains:      Array<{ domain:  string; count: number }>
}

export interface BrowserEntryFilters {
  page?:          number
  page_size?:     number
  search?:        string
  browser?:       string
  artifact_type?: string
  time_from?:     string
  time_to?:       string
  sort_dir?:      string
}

export const browserApi = {
  upload: (caseId: string, file: File) => {
    const fd = new FormData()
    fd.append('file', file)
    return api.post<BrowserFile>(`/browser/${caseId}/upload`, fd, {
      headers: { 'Content-Type': 'multipart/form-data' },
    }).then(r => r.data)
  },

  listFiles: (caseId: string) =>
    api.get<BrowserFile[]>(`/browser/${caseId}/files`).then(r => r.data),

  deleteFile: (caseId: string, fileId: string) =>
    api.delete(`/browser/${caseId}/files/${fileId}`),

  summary: (caseId: string, fileId: string) =>
    api.get<BrowserSummary>(`/browser/${caseId}/files/${fileId}/summary`).then(r => r.data),

  entries: (caseId: string, fileId: string, filters: BrowserEntryFilters = {}) =>
    api.get<BrowserEntriesPage>(`/browser/${caseId}/files/${fileId}/entries`, { params: filters })
       .then(r => r.data),

  reparse: (caseId: string, fileId: string) =>
    api.post<BrowserFile>(`/browser/${caseId}/files/${fileId}/reparse`).then(r => r.data),

  addEvidence: (caseId: string, fileId: string) =>
    api.post<BrowserFile>(`/browser/${caseId}/files/${fileId}/add-evidence`).then(r => r.data),
}

File: backend/app/schemas/browser.py
from __future__ import annotations

import json
from datetime import datetime

from pydantic import BaseModel, computed_field


class BrowserFileOut(BaseModel):
    id:            str
    case_id:       str
    filename:      str
    artifact_type: str
    status:        str
    entry_count:   int | None
    error_msg:     str | None
    uploaded_at:   datetime
    parsed_at:     datetime | None
    added_to_evidence:      bool
    parse_progress:         int
    parse_duration_seconds: int | None
    columns_json:           str | None = None   # raw JSON – parsed below

    model_config = {"from_attributes": True}

    @computed_field
    @property
    def columns(self) -> list[str]:
        """Original CSV column names, decoded from columns_json."""
        if self.columns_json:
            try:
                return json.loads(self.columns_json)
            except Exception:
                return []
        return []


class BrowserEntryOut(BaseModel):
    """One row from a WebX browser-artifact CSV (via DuckDB).

    Normalized fields are the minimal set used for filtering / sorting.
    All original CSV columns are returned verbatim in ``raw_data``.
    """
    row_num:        int
    artifact_type:  str
    event_timestamp: datetime | None
    url:            str | None
    title:          str | None
    browser:        str | None
    profile:        str | None
    username:       str | None
    raw_data:       dict[str, str]   # {original_column_name: value}


class BrowserEntriesPage(BaseModel):
    total:     int
    page:      int
    page_size: int
    pages:     int
    items:     list[BrowserEntryOut]


class BrowserSummary(BaseModel):
    total_entries:    int
    artifact_type:    str
    oldest_timestamp: datetime | None
    newest_timestamp: datetime | None
    top_browsers:     list[dict]   # [{browser, count}]
    top_domains:      list[dict]   # [{domain, count}]
